File: solutions/0214. Shortest Palindrome/0214.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  string shortestPalindrome(string s) {
    long long hashBase = 29;
    long long kMod = 1'000'000'007;
    long long forwHash = 0;
    long long revHash = 0;
    long long powVal = 1;
    int index = -1;

    for (int i = 0; i < s.length(); ++i) {
      char currChar = s[i];
      forwHash = (forwHash * hashBase + (currChar - 'a' + 1)) % kMod;
      revHash = (revHash + (currChar - 'a' + 1) * powVal) % kMod;
      powVal = (powVal * hashBase) % kMod;

      if (forwHash == revHash)
        index = i;
    }

    string suffix = s.substr(index + 1);
    string revSuffix(suffix.rbegin(), suffix.rend());

    return revSuffix + s;
  }
};

File: solutions/0214. Shortest Palindrome/0214.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def shortestPalindrome(self, s: str) -> str:
    hashBase = 29
    kMod = 1_000_000_007
    forwHash = 0
    revHash = 0
    powVal = 1
    index = -1

    for i, currChar in enumerate(s):
      forwHash = (forwHash * hashBase + (ord(currChar) - ord('a') + 1)) % kMod
      revHash = (revHash + (ord(currChar) -  ord('a') + 1) * powVal) % kMod
      powVal = (powVal * hashBase) % kMod

      if forwHash == revHash:
        index = i

    suffix = s[index + 1 :]
    revSuffix = suffix[::-1]

    return revSuffix + s